### PR TITLE
fix(dateparser): add type and validity check

### PR DESCRIPTION
**Pull request by wesleycho**
- Add type and validity check to ensure proper date object is created
- Make code style uniform in scripts

Plunker with fix in action [here](http://plnkr.co/edit/k54L20Ux3VkND7yAK2Kh?p=preview)

This addresses #3713

### CODE REVIEW

**Comment by wesleycho** [comment]
I believe this also fixes #3701

**Comment by rvanbaalen** [comment]
Leaving a comment to keep track on this one. Will review this saturday.

**Comment by chrisirhc** [comment]
@wesleycho , there's a lot of whitespace changes in this PR. Any chance of editing that away? Just want to avoid adding noise to blame.

**Comment by chrisirhc** [comment]
Code changes looks good to me.
I was only wondering if this should log a warning when the base date argument is provided but invalid (`isNaN` check returns true).

**Comment by wesleycho** [comment]
Sure, I can do that sometime later today or tomorrow.

I can add a $log.debug or $log.warn statement about a string where isNaN returns true and falling back to date constructor.

**Comment by wesleycho** [comment]
I reverted all code style changes

File: src/dateparser/dateparser.js [hunk 1]
```diff
@@ -136,6 +136,9 @@ angular.module('ui.bootstrap.dateparser', [])
           milliseconds: baseDate.getMilliseconds()
         };
       } else {
+        if (baseDate) {
+          $log.warn('dateparser:', 'baseDate is a string with no date representation');
```
Comment (chrisirhc) [inline src/dateparser/dateparser.js]: How about `baseDate is not a valid date` ?

Comment (wesleycho) [inline src/dateparser/dateparser.js]: Much better than what I had - changing shortly

**Comment by chrisirhc** [comment]
Hm, I was just testing this, I think this introduces another bug, now there's time added to the string, possibly due to a toISOString conversion that shouldn't be done:

See http://plnkr.co/edit/XCvfhGTYUUraB6mTSVSK?p=preview (there are steps to reproduce on the Plunker)

I'm thinking the change in https://github.com/angular-ui/bootstrap/pull/3294 is needed to actually fix this at the root.

**Comment by wesleycho** [comment]
I tried some experimenting, and it looks like stripping out the `|| new Date(...)` portions from the $viewChangeListener and $parser does not fix this issue at all. In fact, this looks to be a preexisting issue, due to the fact that the view value gets converted to a date as the model value.

**Comment by wesleycho** [comment]
#3643 may also be related.

**Comment by cbosco** [comment]
+1

**Comment by IRus** [comment]
Hi, don't understand this check:

```
if (angular.isDate(baseDate) && !isNaN(baseDate.getTime())) {
```
if baseDate is date getTime always will return an number, right?

**Comment by wesleycho** [comment]
That is not true - try doing `new Date('foo').getTime()`

**Comment by IRus** [comment]
Yep, you right, Thank you for example.

**Comment by wesleycho** [comment]
@chrisirhc what are your thoughts on this? I do not think that the model aspect is fixable, since the model value will always become a date object or `null` currently (even if the `new Date(...)` fallback is stripped).

This can potentially be alleviated if we are able to switch to allowing users to bring in a custom parser & formatter for converting between a user desired model and the view value, which would also open up moment.js support for users, but that is a large & separate task.

I think this PR should make it in in its current incarnation though - it fixes a badly breaking bug in the datepicker.

**Comment by IRus** [comment]
@wesleycho please remove || new Date(...) "fallback" that actually introducing new bugs.

**Comment by wesleycho** [comment]
That is a breaking change that has explicit tests written against it - it is outside the scope of this PR.

**Comment by anilnatha** [comment]
Is this in any way related to Issue #3617?

Reason I ask is because one of the behaviors of the bug illustrated in 3617 is that when you clear a datepicker field, it sets a form's $invalid property to false, even though the field is blank and the field is marked as _not_ required. I created a plunker illustrating the issue and is posted in that thread.

**Comment by llafuente** [comment]
Partially related.
I apply this patch, and the exception is gone, but form is still $invalid.

**Comment by wesleycho** [comment]
@llafuente I am not seeing the form being invalid incorrectly - [here](http://plnkr.co/edit/7HUwgHvV5OfT8qLbpiHR?p=preview) is an updated Plunker.

**Comment by llafuente** [comment]
This patch alone do not solve: #3617

I apply this patch to 0.13, and didn't work, form was $invalid.
I can confirm master + this patch works.

**Comment by wesleycho** [comment]
`master` is the only version that matters. There are a lot of changes since 0.13.0.

**Comment by wesleycho** [comment]
Merging this in shortly, as this check in itself is completely safe & necessary.

**Pull request closed without merging**